File: src/main/java/u04lab/polyglot/minesweeper/model/Cell.java
package u04lab.polyglot.minesweeper.model;

import u04lab.polyglot.minesweeper.Pair;

import java.util.Objects;

public class Cell {

    private final Pair<Integer, Integer> position;

    public Cell(Integer row, Integer column) {
        this.position = new Pair<>(row, column);
    }

    public int getRow() {
        return this.position.getX();
    }

    public int getColumn() {
        return this.position.getY();
    }

    public boolean isAdjacentTo(Cell cell) {
        int x = Math.abs(cell.getRow() - this.getRow());
        int y = Math.abs(cell.getColumn() - this.getColumn());
        return !(x == 0 && y == 0) && x <= 1 && y <= 1;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Cell cell)) return false;
        return position.equals(cell.position);
    }

    @Override
    public int hashCode() {
        return Objects.hash(position);
    }
}

File: src/main/scala/u04lab/polyglot/minesweeper/model/Cell.scala
package u04lab.polyglot.minesweeper.model

case class Cell(row: Int, column: Int):
  private object ComponentDistance: // overkill?
    def unapply(cells: (Cell, Cell)): scala.Option[(Int, Int)] = cells match
      case (a, b) => scala.Some((Math.abs(a.row - b.row), Math.abs(a.column - b.column)))

  def isAdjacentTo(cell: Cell): Boolean = (this, cell) match
    case ComponentDistance(x, y) => !(x == 0 && y == 0) && x <= 1 && y <= 1;
